### PR TITLE
[Merged by Bors] - feat: port Algebra.EuclideanDomain.Defs

### PR DESCRIPTION
**Pull request by zeramorphic**
mathlib3 SHA: f1a2caaf51ef593799107fe9a8d5e411599f3996

### CODE REVIEW

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+  protected remainder : R → R → R
+  /-- The property that links the quotient and remainder functions.
+    This allows us to compute GCDs and LCMs. -/
+  protected quotient_mul_add_remainder_eq : ∀ a b, b * quotient a b + remainder a b = a
+  /-- A well-founded relation on `R`, satisfying `r (a % b) b`.
+    This ensures that the GCD algorithm always terminates. -/
+  protected r : R → R → Prop
+  /-- The relation `r` must be well-founded.
+    This ensures that the GCD algorithm always terminates. -/
+  r_well_founded : WellFounded r
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:89]:
```suggestion
  r_wellFounded : WellFounded r
```

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+theorem div_zero (a : R) : a / 0 = 0 :=
+  EuclideanDomain.quotient_zero a
+#align euclidean_domain.div_zero EuclideanDomain.div_zero
+
+section
+
+open Classical
+
+@[elab_as_elim]
+theorem gcd.induction {P : R → R → Prop} :
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:176]:
```suggestion
theorem GCD.induction {P : R → R → Prop} :
```
The GCD namespace should be in UpperCamelCase

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+      exact a0.symm ▸ H0 b
+    else
+      have _ := mod_lt b a0
+      H1 _ _ a0 (gcd.induction (b % a) a H0 H1)
+  termination_by' ⟨_, r_well_founded⟩
+#align euclidean_domain.gcd.induction EuclideanDomain.gcd.induction
+
+end
+
+section Gcd
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:191]:
```suggestion
section GCD
```

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+/-- An implementation of the extended GCD algorithm.
+At each step we are computing a triple `(r, s, t)`, where `r` is the next value of the GCD
+algorithm, to compute the greatest common divisor of the input (say `x` and `y`), and `s` and `t`
+are the coefficients in front of `x` and `y` to obtain `r` (i.e. `r = s * x + t * y`).
+The function `xgcd_aux` takes in two triples, and from these recursively computes the next triple:
+```
+xgcd_aux (r, s, t) (r', s', t') = xgcd_aux (r' % r, s' - (r' / r) * s, t' - (r' / r) * t) (r, s, t)
+```
+-/
+def xgcd_aux : R → R → R → R → R → R → R × R × R
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:221]: This should be lowerCamelCase, no?

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+#align euclidean_domain.xgcd_aux_rec EuclideanDomain.xgcd_aux_rec
+
+/-- Use the extended GCD algorithm to generate the `a` and `b` values
+  satisfying `gcd x y = x * a + y * b`. -/
+def xgcd (x y : R) : R × R :=
+  (xgcd_aux x 1 0 y 0 1).2
+#align euclidean_domain.xgcd EuclideanDomain.xgcd
+
+/-- The extended GCD `a` value in the equation `gcd x y = x * a + y * b`. -/
+def gcd_a (x y : R) : R :=
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:252]: Also lowerCamelCase.

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+  rw [xgcd, xgcd_zero_left]
+#align euclidean_domain.gcd_b_zero_left EuclideanDomain.gcd_b_zero_left
+
+theorem xgcd_val (x y : R) : xgcd x y = (gcd_a x y, gcd_b x y) :=
+  Prod.mk.eta.symm
+#align euclidean_domain.xgcd_val EuclideanDomain.xgcd_val
+
+end Gcd
+
+section Lcm
```
Comment (rosborn) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:279]:
```suggestion
section LCM
```

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+  `x * a + y * b = gcd a b`.
+* `lcm`: defines the lowest common multiple of two elements `a` and `b` of a Euclidean domain as
+  `a * b / (gcd a b)`
+
+## Main statements
+
+See `Algebra.EuclideanDomain.Basic` for most of the theorems about Eucliean domains,
+including Bézout's lemma.
+
+See `Algebra.EuclideanDomain.Instances` for that facts that `ℤ` is a Euclidean domain,
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:35]: While you're here:
```suggestion
See `Algebra.EuclideanDomain.Instances` for the facts that `ℤ` is a Euclidean domain,
```

File: Mathlib/Algebra/EuclideanDomain/Defs.lean [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+#align euclidean_domain.r EuclideanDomain.r
+#align euclidean_domain.r_well_founded EuclideanDomain.r_wellFounded
+#align euclidean_domain.remainder_lt EuclideanDomain.remainder_lt
+#align euclidean_domain.mul_left_not_lt EuclideanDomain.mul_left_not_lt
+
+namespace EuclideanDomain
+
+variable {R : Type u} [EuclideanDomain R]
+
+-- mathport name: «expr ≺ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/EuclideanDomain/Defs.lean:108]:
```suggestion
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**